rfd/0174-kube-rbac-bootstrap.md

```diff
@@ -0,0 +1,346 @@
+- Product: `@klizhentas` || `@xinding33`
+
+## What
+
+Improve RBAC setup capabilities for Kubernetes and make it easier to use with Teleport by
+adding the ability to define Kubernetes RBAC completely in Teleport and adding the ability to define RBAC setup to be
+automatically provisioned to Kubernetes clusters.
+
+## Why
+
+Currently, users have to rely on third-party tools or manual actions
+when making sure their Kubernetes clusters have the correct roles/role bindings and that those are in sync with the RBAC setup defined in Teleport.
+The changes we propose will make it easier for clients to start using Teleport for accessing Kubernetes clusters.
+
+## Scope
+
+This RFD is focused only on Kubernetes resources related to RBAC functionality, other types of resources are out of scope
+and we will not allow them to be provisioned. Although other resources might be added in the future.
+
+## Details
+
+### Improving Day 1 experience - Adding capability to manage Kubernetes RBAC from Teleport. 
+
+Currently users rely on `kubernetes_groups` and/or `kubernetes_users` fields of the role definition when setting up
+access for Kubernetes. This requires roles/bindings to be created by the client in Kubernetes cluster independently from the 
+RBAC setup in the Teleport. These roles/bindings also should be kept in sync with the Teleport RBAC side,
+which creates additional hurdles for clients, especially the new ones, when using Kubernetes access. 
+To improve this situation we will introduce auto-provisioning of the Kubernetes RBAC based on 
+the Teleport roles, which means users will be able to define required permissions completely in the Teleport, and under the hood
+we will keep them in sync transparently for the users.
+
+We will add a new field to the role definition - `kubernetes_permissions`. That field will contain desired Kubernetes RBAC permissions
+for that role and will be similar in structure to the native Kubernetes RBAC:
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: kube-role
+spec:
+  allow:
+    kubernetes_labels:
+      "*": "*"
+    kubernetes_groups: ["kube_group"]
+    kubernetes_permissions:
+      namespaces: ["namespace1", "namespace2"]
+      rules:
+      - resources: ["pod", "pod/exec"]
+        verbs: ["get", "list", "create"]
+      - resources: ["deployments"]
+        apiGroups: ["app"]
+        verbs: ["*"]
+      - resources: ["secrets"]
+        apiGroups: [""]
+        resourceNames: ["secret1", "secret2"]
+        verbs: ["get", "list""]
+```
+
+Presence of this field will signal that user wants Teleport to auto-provision these permissions into the Kubernetes clusters.
+Teleport will translate permissions defined in that field into Kubernetes RBAC definitions and will automatically create roles and bindings
+on the target Kubernetes clusters (as specified by the labels). If `namespaces` field includes '*' Teleport will create ClusterRole/ClusterRoleBinding
```

Comment:
>If `namespaces` field includes '*' Teleport will create ClusterRole/ClusterRoleBinding

I believe that using `*` should not imply ClusterRole/ClusterRoleBinding. First, some resources are cluster-wide and need to be distinguished from resources that can be namespaced. Second, `*` typically means all namespaces, but `some-team-*` only matches a subset but both include a matcher.

Therefore, from my perspective, `namespaces: ['*']` should indicate creating a role and rolebinding in every namespace. If `namespaces` is omitted, it should signify the creation of a clusterrole.

Reply:
> creating a role and rolebinding in every namespace

Oh no, namespace watcher and race-conditions incoming. (I am a users that can create namespaces and create resources in the namespace `*`. If I `kubectl create ns foo || kubectl apply -f manifest.yaml -n foo` the second command will likely fail as the RBAC is not there yet.

Reply:
> Oh no, namespace watcher and race-conditions incoming.

@hugoShaka, you will always need to monitor namespaces, and the issues will persist. There's no guarantee that the namespace specified in `kubernetes_permissions` will exist at any given moment, but it may exist at a later time. Therefore, a namespace watcher is essential, and we must reconcile based on that, even if you don't use the wildcard.

If your role specifies `namespace: abc` and you run `kubectl create ns abc || kubectl apply ...`, you will encounter the same problem. For me, translating `namespace: *` into a cluster role is problematic because it doesn’t align well with specifying cluster-wide resources. Cluster-wide resources don’t have a namespace, so this approach seems inconsistent.